### PR TITLE
JAMES-2516:Created a SenderDomainId && RecipeintDomainIs matchers

### PR DESCRIPTION
**Pull request by athulyaraj**
#1834

### CODE REVIEW

**Comment by athulyaraj** [comment]
@chibenwa I have addressed the review comments given. Please review.

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SenderDomainIsTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                        "james.apache.org, james3.apache.org, james2.apache.org, james4.apache.org, james5.apache.org")
+                .build());
+
+        FakeMail fakeMail = FakeMail.builder().recipient(recipient).sender("other@james7.apache.org").build();
+
+        assertThat(matcher.match(fakeMail)).isEmpty();
+    }
+
+    @Test
+    void shouldNotMatchWhenNullSenderDomain() throws Exception {
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SenderDomainIsTest.java]: This test is `shouldNotMatchWhenNoSenderDomain`

An other meaningfull test would be: `FakeMail.builder().sender(MailAddress.nullSender())...`

Comment (athulyaraj) [reply mailet/standard/src/test/java/org/apache/james/transport/matchers/SenderDomainIsTest.java]: Added these changes as well.

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+
+/**
+ * This will return recipients mathcing a configured domain.
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java]: s/mathcing/matching/

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ * &lt;mailet match=&quot;RecipientDomainIs=&lt;domain.com&gt;&quot; class=&quot;&lt;any-class&gt;&quot;&gt;
+ * </CODE>
+ * </PRE>
+ * @version CVS $Revision$ $Date$
+ * @since 3.2.0
+ */
+public class RecipientDomainIs extends GenericRecipientMatcher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RecipientDomainIs.class);
+
+    private Collection<Domain> receipientDomains;
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java]: s/receipientDomains/recipientDomains/

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    public void init() {
+        String condition = getCondition();
+        Preconditions.checkNotNull(condition, "'condition' should not be null");
+
+        receipientDomains = parseDomainsList(condition);
+    }
+
+    @VisibleForTesting
+    Collection<Domain> parseDomainsList(String condition) {
+        return Splitter.onPattern("(, |,| )").omitEmptyStrings().splitToList(condition).stream().map(Domain::of)
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java]: can you please add some carriage returns to help reading this please?
```suggestion
        return Splitter.onPattern("(, |,| )")
            .omitEmptyStrings()
            .splitToList(condition)
            .stream()
            .map(Domain::of)
```

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                .collect(Guavate.toImmutableList());
+    }
+
+    @Override
+    public boolean matchRecipient(MailAddress recipient) {
+        try {
+            return receipientDomains.contains(recipient.getDomain());
+
+        } catch (Exception ex) {
+            LOGGER.error("Exception happened while finding receipient domain match", ex);
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java]: also recipient :)

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/SenderDomainIs.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        String condition = getCondition();
+        Preconditions.checkNotNull(condition, "'condition' should not be null");
+
+        senderDomains = parseDomainsList(condition);
+    }
+
+    @VisibleForTesting
+    Collection<Domain> parseDomainsList(String condition) {
+        return Splitter.onPattern("(, |,| )").omitEmptyStrings().splitToList(condition).stream().map(Domain::of)
+                .collect(Guavate.toImmutableList());
```
Comment (rouazana) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/SenderDomainIs.java]: same, can you add some carriage return please?

File: mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    @VisibleForTesting
+    Collection<Domain> parseDomainsList(String condition) {
+        return Splitter.onPattern("(, |,| )").omitEmptyStrings().splitToList(condition).stream().map(Domain::of)
+                .collect(Guavate.toImmutableList());
+    }
+
+    @Override
+    public boolean matchRecipient(MailAddress recipient) {
+        try {
+            return receipientDomains.contains(recipient.getDomain());
```
Comment (trantienduchn) [inline mailet/standard/src/main/java/org/apache/james/transport/matchers/RecipientDomainIs.java]: forgive me but I don't understand your intention while putting this statement in try catch block. These method callings seem quite hard to throw an exception. except when recipeient is null

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/RecipientDomainIsTest.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+        FakeMail fakeMail = FakeMail.builder().recipients(OTHER_AT_JAMES2, ANY_AT_JAMES2).build();
+
+        assertThat(matcher.match(fakeMail)).isEmpty();
+    }
+
+    @Test
+    void initShouldThrowOnMissingCondition() {
+        assertThatThrownBy(() -> matcher.init(FakeMatcherConfig.builder().matcherName("RecipientDomainIs").build()))
+                .isInstanceOf(NullPointerException.class);
```
Comment (trantienduchn) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/RecipientDomainIsTest.java]: indentations of your classes should be take care of. We use one level of indentation(4 spaces) when doing statement line break.

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SenderDomainIsTest.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.apache.james.core.MailAddress;
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.FakeMatcherConfig;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class SenderDomainIsTest {
```
Comment (trantienduchn) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SenderDomainIsTest.java]: class could goes with default modifier :)

**Comment by athulyaraj** [comment]
Addressed latest review comments

**Comment by trantienduchn** [comment]
```
org.apache.james.mpt.imapmailbox.external.james.JamesDeploymentValidation
[2c3f4ab5223a50acb4f614dacd6207ee91e7aa15] [ERROR] validateDeployment(org.apache.james.mpt.imapmailbox.external.james.JamesDeploymentValidation)  Time elapsed: 9.737 s  <<< ERROR!
```
Related to old bug, we fixed this issues few days ago, could you rebase your branch on master and push?

**Comment by athulyaraj** [comment]
i merged latest changes in master to my branch

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/RecipientDomainIsTest.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import javax.mail.MessagingException;
+
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.FakeMatcherConfig;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+ class RecipientDomainIsTest {
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/RecipientDomainIsTest.java]: unnecessary space at the beginning of the line

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/RecipientDomainIsTest.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        matcher = new RecipientDomainIs();
+    }
+
+    @Test
+    void shouldMatchOneAddress() throws MessagingException {
+        matcher.init(
+                FakeMatcherConfig.builder()
+                .matcherName("RecipientDomainIs")
+                .condition("james.apache.org")
+                .build());
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/RecipientDomainIsTest.java]: still a small indent problem, should be:
```suggestion
        matcher.init(
            FakeMatcherConfig.builder()
                .matcherName("RecipientDomainIs")
                .condition("james.apache.org")
                .build());
```
idem later in this file

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/SenderDomainIsTest.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.apache.james.core.MailAddress;
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.FakeMatcherConfig;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+ class SenderDomainIsTest {
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/SenderDomainIsTest.java]: also a space at the beginning

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/RecipientDomainIsTest.java (hunk 10 above)
Comment (rouazana) [reply mailet/standard/src/test/java/org/apache/james/transport/matchers/RecipientDomainIsTest.java]: this comment has not been solved. Here github's tool should should, let's try it:
```suggestion
class RecipientDomainIsTest {
```

**Comment by chibenwa** [comment]
Hi,

This pull request has just been merged.

Thank you again for your dedication toward the James project. If you need help, want to [contribute](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20newbie%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC) [documentation](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20documentation%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), new [features](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20feature%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), tests or [bugfix](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20easyfix%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), please don't hesitate to reach us.

Thanks again,

Cheers,

Benoit

**Comment by chibenwa** [comment]
Hi,

This pull request has just been merged.

Thank you again for your dedication toward the James project. If you need help, want to [contribute](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20newbie%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC) [documentation](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20documentation%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), new [features](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20feature%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), tests or [bugfix](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20easyfix%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), please don't hesitate to reach us.

Thanks again,

Cheers,

Benoit

**Pull request closed without merging**